src/varname.jl

```diff
@@ -26,29 +29,42 @@ julia> @varname x[:, 1][1+1]
 x[:,1][2]
 ```
 """
-struct VarName{sym, T<:Tuple}
+struct VarName{sym, T<:Lens}
     indexing::T
 
-    VarName{sym}(indexing::Tuple=()) where {sym} = new{sym,typeof(indexing)}(indexing)
+    VarName{sym}(indexing=IdentityLens()) where {sym} = new{sym,typeof(indexing)}(indexing)
 end
 
+# A bit of backwards compatibility.
+# TODO: Should we deprecate this?
+VarName{sym}(indexing::Tuple) where {sym} = VarName{sym}(tuple2indexlens(indexing))
+
 """
-    VarName(vn::VarName, indexing=())
+    VarName(vn::VarName, indexing::Lens)
+    VarName(vn::VarName, indexing::Tuple)
 
 Return a copy of `vn` with a new index `indexing`.
 
-```jldoctest
+```jldoctest; setup=:(using Setfield)
+julia> VarName(@varname(x[1][2:3]), Setfield.IndexLens((2,)))
+x[2]
+
 julia> VarName(@varname(x[1][2:3]), ((2,),))
 x[2]
 
 julia> VarName(@varname(x[1][2:3]))
 x
 ```
 """
-function VarName(vn::VarName, indexing::Tuple = ())
-    return VarName{getsym(vn)}(indexing)
+VarName(vn::VarName, indexing::Lens=IdentityLens()) = VarName{getsym(vn)}(indexing)
+
+function VarName(vn::VarName, indexing::Tuple)
+    return VarName{getsym(vn)}(tuple2indexlens(indexing))
 end
 
+tuple2indexlens(indexing::Tuple{}) = IdentityLens()
+tuple2indexlens(indexing::Tuple{<:Tuple}) = IndexLens(first(indexing))
+tuple2indexlens(indexing::Tuple) = IndexLens(first(indexing)) ∘ tuple2indexlens(indexing[2:end])
```

Comment:
```suggestion
tuple2indexlens(indexing::Tuple) = mapfoldr(IndexLens, ∘, indexing, init=IdentityLens())
```

Reply:
Also, `2indexlens` -- isn't that a bit misleading, as the function does not return an index lens except in one special case (or never, when implemented as a fold)?  Maybe `tuple2indexlenses`?

Reply:
(OK, `compose` is actualy [doing the intelligent thing](https://github.com/jw3126/Setfield.jl/blob/3ce508dfaf680e7aa31997b61568a75913a69ea7/src/lens.jl#L132) and discards `IdentityLens`.)

Reply:
Maybe `indextuple2lens`?

Reply:
Btw, reason why I don't use `mapreduce` here is just because I'm more confident that the compiler can handle the recursion rather than in `mapreduce`. I've found `mapreduce` to be a bit unreliable when it comes to type inference, e.g. the `find_indexing` or whatever it's called in DynamicPPL that I discovered a couple of weeks ago.

The renaming I'm down for though. Happy with `indextuple2lens`?